clang-tools-extra/clang-tidy/hooks/GuardInForCheck.cpp

```diff
@@ -19,116 +19,280 @@ namespace hooks {
 namespace {
 
 const int64_t MAX_FOR_LIMIT = 10000;
+const int MAX_NESTED_LOOP_LEVEL = 10;
 
 bool areSameVariable(const ValueDecl *First, const ValueDecl *Second) {
   return First && Second &&
     First->getCanonicalDecl() == Second->getCanonicalDecl();
 }
 
+auto isUnguarded()  {
+  return unless(has(binaryOperator(has(callExpr(callee(functionDecl(hasName("_g"))))))));
 }
 
-void GuardInForCheck::registerMatchers(MatchFinder *Finder) {
-  // based on https://clang.llvm.org/docs/LibASTMatchersTutorial.html
-  const auto CondVarNameExpr = ignoringParenImpCasts(declRefExpr(
-          to(varDecl(hasType(isInteger())).bind("condVarName"))));
+std::string buildStr(std::string str, int nestingLevel) {
+  return str + std::to_string(nestingLevel);
+}
+
+auto conditionLimit(int nestingLevel) {
+  return expr().bind(buildStr("constLimit-", nestingLevel));
+}
+
+auto conditionVarName(int nestingLevel) {
+  return ignoringParenImpCasts(declRefExpr(
+          to(varDecl(hasType(isInteger())).bind(buildStr("condVarName-", nestingLevel)))));
+}
+
+auto lessThanCondition(int nestingLevel) {
+  return binaryOperator(anyOf(hasOperatorName("<"), hasOperatorName("<=")),    //e.g. i < 2 or 2 < i
+            hasLHS(anyOf(conditionVarName(nestingLevel), conditionLimit(nestingLevel))),
+            hasRHS(anyOf(conditionVarName(nestingLevel), conditionLimit(nestingLevel)))
+          ).bind(buildStr("condOp-", nestingLevel));
+}
+
+auto greaterThanCondition(int nestingLevel) {
+  return binaryOperator(anyOf(hasOperatorName(">"), hasOperatorName(">=")),   //e.g. 2 > i or i > 2
+            hasLHS(anyOf(conditionVarName(nestingLevel), conditionLimit(nestingLevel))),
+            hasRHS(anyOf(conditionVarName(nestingLevel), conditionLimit(nestingLevel)))
+          ).bind(buildStr("condOp-", nestingLevel));
+}
+
+auto notEqualCondition(int nestingLevel) {
+  return binaryOperator(hasOperatorName("!="),    //e.g. i != 2 or 2 != i
+            hasLHS(anyOf(conditionVarName(nestingLevel), conditionLimit(nestingLevel))),
+            hasRHS(anyOf(conditionVarName(nestingLevel), conditionLimit(nestingLevel)))).bind(buildStr("condOp-", nestingLevel));
+}
+
+auto conditionHasGuard(int nestingLevel) {
+  return binaryOperator(hasOperatorName(","),
+    hasEitherOperand(callExpr(callee(functionDecl(
```

Comment:
it probably doesn't make a practical difference, but shouldn't it be `hasLeftOperand`?

Reply:
With `hasEitherOperand` the `GUARD(x)` call can be on either left or right side of the operator - both cases will be matched.

Reply:
> With `hasEitherOperand` the `GUARD(x)` call can be on either left or right side of the operator - both cases will be matched.

Yes, and that's the problem - `GUARD` on the right side of the operator is incorrect:
```
for (int i = 0; i < 3, GUARD(3); ++i)
```
will (try to) iterate forever, because the result of the comma expression is the return value of `_g`, i.e. always true...

Reply:
Ah right!